src/olympia/versions/models.py

```diff
@@ -1423,6 +1438,103 @@ def __str__(self):
         except ObjectDoesNotExist:
             return pretty_application
 
+    def version_range_contains_forbidden_compatibility(self):
+        """
+        Return whether the instance contains forbidden ranges.
+
+        Used for Android to disallow non line/recommended to be compatible with
+        Fenix before General Availability.
+        """
+        if self.application != amo.ANDROID.id:
+            return False
+
+        addon = self.version.addon
+
+        if addon.type != amo.ADDON_EXTENSION:
+            # We don't care about non-extension compatibility for this.
+            return False
+
+        # Recommended/line for Android are allowed to set compatibility in the
+        # limited range.
+        if (
+            addon.promoted
+            and addon.promoted.group in (RECOMMENDED, LINE)
+            and amo.ANDROID in addon.promoted.approved_applications
+        ):
+            return False
+
+        is_min_in_limited_range = self.min.version_int >= version_int(
+            self.ANDROID_LIMITED_COMPATIBILITY_RANGE[0]
+        ) and self.min.version_int < version_int(
+            self.ANDROID_LIMITED_COMPATIBILITY_RANGE[1]
+        )
+        is_max_in_limited_range = self.max.version_int >= version_int(
+            self.ANDROID_LIMITED_COMPATIBILITY_RANGE[0]
+        ) and self.max.version_int < version_int(
+            self.ANDROID_LIMITED_COMPATIBILITY_RANGE[1]
+        )
+        is_range_bigger_than_limited_range = self.min.version_int < version_int(
+            self.ANDROID_LIMITED_COMPATIBILITY_RANGE[0]
+        ) and self.max.version_int >= version_int(
+            self.ANDROID_LIMITED_COMPATIBILITY_RANGE[1]
+        )
+        return (
+            is_min_in_limited_range
+            or is_max_in_limited_range
+            or is_range_bigger_than_limited_range
+        )
```

Comment:
I pushed e10591b5167a33d441a0065e6f52f449886fa74b to make things easier to read.

The 3 different cases are important - ultimately that's what inside the range that matters.

Reply:
The code suggestion is (supposed to be) a replacement for all three of the different cases - albeit written in a non-obvious way.

Reply:
Ah, what confused me is that the first condition in your suggestion should be a `>` (or we should reverse the ordering and use `<`) then. Otherwise it would not allow `119.0a1` as the `min`.

I'm not sure which is better tbh.

Reply:
Ended up using your suggestion with that fix and a couple more more tests in ef6b2428d4a17e84d3bd16cbaef0a4eeff5d18b8